### PR TITLE
Icon Grid Block

### PR DESCRIPTION
**Pull request by ScottGarrison**
## What does this do/fix?

This PR adds in the Icon Grid Block into SQ1.

Nothing too major to call out here as it's pretty straight forward. 
Per DaveE's previous feedback on other blocks, I added the Lead In and CTA to the Content Block to keep things consistent.

image

### CODE REVIEW

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+					'type'  => 'link',
+				] )
+			)->add_field(
+				$this->get_icon_section()
+			);
+
+		//==========================================
+		// Setting Fields
+		//==========================================
+		$this->add_section( new Field_Section( self::SECTION_SETTINGS, __( 'Settings', 'tribe' ), 'accordion' ) );
```
Comment (msawicki) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:79]: If there aren't any settings can we just delete this?

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		$group->set_attributes( [
+			'label'        => __( 'Icon Section', 'tribe' ),
+			'name'         => self::ICONS,
+			'layout'       => 'block',
+			'min'          => 0,
+			'max'          => 12,
+			'button_label' => 'Add Icon Section',
+		] );
+
+		$icon_image = new Field( self::ICON_IMAGE, [
```
Comment (msawicki) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:96]: You don't have to make any of these changes but we've made some improvements to registering fields that are kind of handy and nice for code clutter.

```
$group = new Repeater( self::NAME . '_' . self::ICONS, [
	'label'        => __( 'Icon Section', 'tribe' ),
	'name'         => self::ICONS,
	'layout'       => 'block',
	'min'          => 0,
	'max'          => 12,
	'button_label' => 'Add Icon Section',
] );

$group->add_field(
    new Field( self::ICON_IMAGE, [
		'label'         => __( 'Icon Image', 'tribe' ),
		'name'          => self::ICON_IMAGE,
		'type'          => 'image',
		'return_format' => 'id',
		'preview_size'  => 'medium',
		'instructions'  => __( 'Recommended image size: 100px wide with any aspect ratio.', 'tribe' ),
	] )
 )->add_field(
    new Field( self::ICON_TITLE, [
		'label' => __( 'Icon Title', 'tribe' ),
		'name'  => self::ICON_TITLE,
		'type'  => 'text',
	] )
)->add_field(
    new Field( self::ICON_DESCRIPTION, [
		'label' => __( 'Icon Description', 'tribe' ),
		'name'  => self::ICON_DESCRIPTION,
		'type'  => 'wysiwyg',
     ] )
)->add_field(
    new Field( self::ICON_LINK, [
		'label' => __( 'Icon Section Link', 'tribe' ),
		'name'  => self::ICON_LINK,
		'type'  => 'link',
    ] )
);

```
We aren't very consistent here _yet_ so again, no change needed but FYI these are available.

**Review by msawicki — APPROVED** [review]

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+					'components/image/image',
+					null,
+					[
+						Image_Controller::IMG_ID       => $card['image'] ?? null,
+						Image_Controller::AS_BG        => false,
+						Image_Controller::SRC_SIZE     => 'medium_large',
+						Image_Controller::SRCSET_SIZES => [
+							'medium',
+							'medium_large',
+						],
```
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:234]: Same question here re: these image sizes. Are these actual sizes in WordPress?

**Review by dpellenwood — APPROVED** [review]
Couple questions, but this looks great, @ScottGarrison . Thank you for tackling this!

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php (hunk 3 above)
Comment (ScottGarrison) [reply wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:234]: It's pretty hard to find the documentation for it but it's listed here as an option: https://codex.wordpress.org/Post_Thumbnails#Thumbnail_Sizes

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	public const LEADIN      = 'leadin';
+	public const CTA         = 'cta';
+
+	public const ICONS            = 'icons';
+	public const ICON_IMAGE       = 'image';
+	public const ICON_TITLE       = 'icon_title';
+	public const ICON_DESCRIPTION = 'icon_description';
+	public const ICON_LINK        = 'link';
+
+	public function add_block() {
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:29]:
```suggestion
	public function add_block(): void {
```

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+				'anchor' => true,
+				'html'   => false,
+			],
+		] ) );
+	}
+
+	/**
+	 * Register Fields for block
+	 */
+	public function add_fields() {
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:47]:
```suggestion
	public function add_fields(): void {
```

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+				] )
+			)->add_field(
+				$this->get_icon_section()
+			);
+	}
+
+	/**
+	 * @return Repeater
+	 */
+	protected function get_icon_section() {
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:84]:
```suggestion
	protected function get_icon_section(): Repeater {
```

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	public const SECTION_CONTENT  = 's-content';
+	public const SECTION_SETTINGS = 's-settings';
+
+	public const TITLE       = 'title';
+	public const DESCRIPTION = 'description';
+	public const LEADIN      = 'leadin';
+	public const CTA         = 'cta';
+
+	public const ICONS            = 'icons';
+	public const ICON_IMAGE       = 'image';
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:24]: Prefix this with `icon_` as well?

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	public const ICON_TITLE       = 'icon_title';
+	public const ICON_DESCRIPTION = 'icon_description';
+	public const ICON_LINK        = 'link';
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:27]: Prefix this with `icon_`?

File: wp-content/themes/core/blocks/icongrid/icongrid.php [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<?php
+declare( strict_types=1 );
```
Comment (defunctl) [inline wp-content/themes/core/blocks/icongrid/icongrid.php:2]: Move this to top line in each file, will make future PHPCS stuff easier for sure.

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+			Link_Controller::CLASSES => [
+				'c-block__cta-link',
+				'a-btn',
+				'a-btn--has-icon-after', 
+				'icon-arrow-right'
+			],
+		] );
+	}
+
+	public function get_icon_card_args() {
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:195]:
```suggestion
	public function get_icon_card_args(): array {
```

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+		if ( empty( $this->icons ) ) {
+			return [];
+		}
+
+		$cards = [];
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:200]: Can move this above the empty check and then just `return $cards;` if `$this->icons` is empty.

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	 * @return Repeater
+	 */
+	protected function get_icon_section() {
+		$group = new Repeater( self::NAME . '_' . self::ICONS, [
+			'label'        => __( 'Icon Section', 'tribe' ),
+			'name'         => self::ICONS,
+			'layout'       => 'block',
+			'min'          => 0,
+			'max'          => 12,
+			'button_label' => 'Add Icon Section',
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:91]:
```suggestion
			'button_label' => __( 'Add Icon Section', 'tribe' ),
```

**Review by defunctl — CHANGES_REQUESTED** [review]

**Comment by defunctl** [comment]
@ScottGarrison you'll want to fix the PHPCS errors found [here](https://github.com/moderntribe/square-one/pull/674/checks?check_run_id=1832112798) as well.

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	public const ICONS            = 'icons';
+	public const ICON_IMAGE       = 'icon_image';
+	public const ICON_TITLE       = 'icon_title';
+	public const ICON_DESCRIPTION = 'icon_description';
+	public const ICON_LINK        = 'icon_link';
+
+	public function add_block(): void {
+		$this->set_block( new Block( self::NAME, [
+			'title'       => __( 'Icon Grid', 'tribe' ),
+			'description' => __( 'A grid layout block with icon selections', 'tribe' ), // TODO: describe the block
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:31]: Can remove any TODO's that are done here.

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+namespace Tribe\Project\Blocks\Types\Icon_Grid;
+
+use Tribe\Libs\ACF\Block;
+use Tribe\Libs\ACF\Block_Config;
+use Tribe\Libs\ACF\Field;
+use Tribe\Libs\ACF\Field_Section;
+use Tribe\Libs\ACF\Repeater;
+
+class Icon_Grid extends Block_Config {
+	public const NAME = 'icongrid';
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid.php:12]:
```suggestion
	
	public const NAME = 'icongrid';
```

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid_Model.php [hunk 15]
```diff
@@ -0,0 +1,44 @@
+<?php
+declare( strict_types=1 );
```
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid_Model.php:2]: can move this to the top line

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+		$this->content_classes[] = 'g-centered';
+
+		return Markup_Utils::class_attribute( $this->content_classes );
+	}
+
+	/**
+	 * @return array
+	 */
+	public function get_header_args(): array {
+		if ( empty( $this->title ) && empty( $this->description ) ) {
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:109]: Both title and description have to be empty for this not to proceed, that's the intention right?

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+	 * @return Deferred_Component
+	 */
+	private function get_leadin(): Deferred_Component {
+		return defer_template_part( 'components/text/text', null, [
+			Text_Controller::CLASSES => [
+				'c-block__leadin',
+				'b-icon-grid__leadin',
+				'h6',
+			],
+			Text_Controller::CONTENT => $this->leadin ?? '',
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:138]:
```suggestion
			Text_Controller::CONTENT => $this->leadin,
```
You're setting defaults in the constructor, so the `??` operator isn't needed

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+	 * @return Deferred_Component
+	 */
+	private function get_title(): Deferred_Component {
+		return defer_template_part( 'components/text/text', null, [
+			Text_Controller::TAG     => 'h2',
+			Text_Controller::CLASSES => [
+				'c-block__title',
+				'h3',
+			],
+			Text_Controller::CONTENT => $this->title ?? '',
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:152]:
```suggestion
			Text_Controller::CONTENT => $this->title,
```

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+	 */
+	private function get_content(): Deferred_Component {
+		return defer_template_part( 'components/container/container', null, [
+			Container_Controller::CLASSES => [
+				'c-block__description',
+				'b-icon-grid__description',
+				't-sink',
+				's-sink',
+			],
+			Container_Controller::CONTENT => $this->description ?? '',
```
Comment (defunctl) [inline wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:167]:
```suggestion
			Container_Controller::CONTENT => $this->description,
```

**Review by defunctl — CHANGES_REQUESTED** [review]
A couple of changes/questions but nice job.

File: wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php (hunk 16 above)
Comment (ScottGarrison) [reply wp-content/themes/core/components/blocks/icon_grid/Icon_Grid_Controller.php:109]: Correct. That's how we have all the other blocks set up currently.

**Comment by ScottGarrison** [comment]
@defunctl Updated the recent requests

**Comment by ryanurban** [comment]
Go ahead and merge this @ScottGarrison ❤️

File: wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid_Model.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Class Icon_Grid_Model
+ *
+ * Responsible for mapping values from the block to arguments
+ * for the component
+ */
+class Icon_Grid_Model extends Base_Model {
+	public function get_data(): array {
+		return [
+			Icon_Grid_Controller::CLASSES     => $this->get_classes(),
```
Comment (ryanurban) [inline wp-content/plugins/core/src/Blocks/Types/Icon_Grid/Icon_Grid_Model.php:19]: Just caught this in porting,  need to add: `Icon_Grid_Controller::ATTRS => $this->get_attrs(),` so we get any block ID / class data.

**Pull request merged**